man/tomdoc.5
.\" generated with Ronn/v0.5
.\" http://github.com/rtomayko/ronn/
.
.TH "TOMDOC" "5" "May 2010" "MOJOMBO" "TomDoc Manual"
.
.SH "NAME"
\fBtomdoc\fR \-\- TomDoc for Ruby \- Version 0.9.0
.
.SH "Purpose"
TomDoc is a code documentation specification that helps you write precise
documentation that is nice to read in plain text, yet structured enough to be
automatically extracted and processed by a machine.
.
.P
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
interpreted as described in RFC 2119.
.
.SH "Class/Module Documentation"
TomDoc for classes and modules consists of a block of single comment markers
(#) that appear directly above the class/module definition. Lines SHOULD be
wrapped at 80 characters. Lines that contain text MUST be separated from the
comment marker by a single space. Lines that do not contain text SHOULD
consist of just a comment marker (no trailing spaces).
.
.P
Code examples SHOULD be indented two spaces (three spaces from the comment
marker).
.
.IP "" 4
.
.nf

# Various methods useful for performing mathematical operations. All
# methods are module methods and should be called on the Math module.
# For example:
#
#   Math.square_root(9)
#   # => 3
#
module Math
  ...
end
.
.fi
.
.IP "" 0
.
.SH "Method Documentation"
A quick example will serve to best illustrate the TomDoc method documentation
format:
.
.IP "" 4
.
.nf

# Duplicate some text an abitrary number of times.
#
# text  \- The String to be duplicated.
# count \- The Integer number of times to duplicate the text.
#
# Examples
#
#   multiplex('Tom', 4)
#   # => 'TomTomTomTom'
#
# Returns the duplicated String.
def multiplex(text, count)
  text * count
end
.
.fi
.
.IP "" 0
.
.P
TomDoc for a specific method consists of a block of single comment markers (#)
that appears directly above the method. There MUST NOT be a blank line between
the comment block and the method definition. A TomDoc method block consists of
a description section (required), an arguments section (required if the method
takes any arguments), an examples section (optional), and a returns section
(required). Lines that contain text MUST be separated from the comment
marker by a single space. Lines that do not contain text SHOULD consist of
just a comment marker (no trailing spaces).
.
.SS "The Description Section"
The description section SHOULD be in plain sentences. Each sentence SHOULD end
with a period. Good descriptions explain what the code does at a high level.
Make sure to explain any unexpected behavior that the method may have, or any
pitfalls that the user may experience. Lines SHOULD be wrapped at 80
characters.
.
.P
If a method's description begins with "Public:" then that method will be
considered part of the project's public API. For example:
.
.IP "" 4
.
.nf

# Public: Initialize a new Widget.
.
.fi
.
.IP "" 0
.
.P
This annotation is designed to let developers know which methods are
considered stable. You SHOULD use this to document the public API of your
project. This information can then be used along with \fISemantic
Versioning\fR to inform decisions on when major, minor, and
patch versions should be incremented.
.
.P
If a method's description begins with "Deprecated:" then that method will be
considered as deprecated and users will know that it will be removed in a
future version.
.
.SS "The Arguments Section"
The arguments section consists of a list of arguments. Each list item MUST be
comprised of the name of the argument, a dash, and an explanation of the
argument in plain sentences. The expected type (or types) of each argument
SHOULD be clearly indicated in the explanation. When you specify a type, use
the proper classname of the type (for instance, use 'String' instead of
'string' to refer to a String type). The dashes following each argument name
should be lined up in a single column. Lines SHOULD be wrapped at 80 columns.
If an explanation is longer than that, additional lines MUST be indented at
least two spaces but SHOULD be indented to match the indentation of the
explanation. For example:
.
.IP "" 4
.
.nf

# element \- The Symbol representation of the element. The Symbol should
#           contain only lowercase ASCII alpha characters.
.
.fi
.
.IP "" 0
.
.P
All arguments are assumed to be required. If an argument is optional, you MUST
specify the default value:
.
.IP "" 4
.
.nf

# host \- The String hostname to bind (default: '0.0.0.0').
.
.fi
.
.IP "" 0
.
.P
For hash arguments, you SHOULD enumerate each valid option in a way similar
to how normal arguments are defined:
.
.IP "" 4
.
.nf

# options \- The Hash options used to refine the selection (default: {}):
#           :color  \- The String color to restrict by (optional).
#           :weight \- The Float weight to restrict by. The weight should
#                     be specified in grams (optional).
.
.fi
.
.IP "" 0
.
.SS "The Examples Section"
The examples section MUST start with the word "Examples" on a line by
itself. The next line SHOULD be blank. The following lines SHOULD be indented
by two spaces (three spaces from the initial comment marker) and contain code
that shows off how to call the method and (optional) examples of what it
returns. Everything under the "Examples" line should be considered code, so
make sure you comment out lines that show return values. Separate examples
should be separated by a blank line. For example:
.
.IP "" 4
.
.nf

# Examples
#
#   multiplex('x', 4)
#   # => 'xxxx'
#
#   multiplex('apple', 2)
#   # => 'appleapple'
.
.fi
.
.IP "" 0
.
.SS "The Returns Section"
The returns section should explain in plain sentences what is returned from
the method. The line MUST begin with "Returns". If only a single thing is
returned, state the nature and type of the value. For example:
.
.IP "" 4
.
.nf

# Returns the duplicated String.
.
.fi
.
.IP "" 0
.
.P
If several different types may be returned, list all of them. For example:
.
.IP "" 4
.
.nf

# Returns the given element Symbol or nil if none was found.
.
.fi
.
.IP "" 0
.
.P
If the return value of the method is not intended to be used, then you should
simply state:
.
.IP "" 4
.
.nf

# Returns nothing.
.
.fi
.
.IP "" 0
.
.P
If the method raises exceptions that the caller may be interested in, add
additional lines that explain each exception and under what conditions it may
be encountered. The lines MUST begin with "Raises". For example:
.
.IP "" 4
.
.nf

# Returns nothing.
# Raises Errno::ENOENT if the file cannot be found.
# Raises Errno::EACCES if the file cannot be accessed.
.
.fi
.
.IP "" 0
.
.P
Lines SHOULD be wrapped at 80 columns. Wrapped lines MUST be indented under
the above line by at least two spaces. For example:
.
.IP "" 4
.
.nf

# Returns the atomic mass of the element as a Float. The value is in
#   unified atomic mass units.
.
.fi
.
.IP "" 0
.
.SH "Special Considerations"
.
.SS "Attributes"
Ruby's built in \fBattr_reader\fR, \fBattr_writer\fR, and \fBattr_accessor\fR require a
bit more consideration. With TomDoc you SHOULD NOT use \fBattr_access\fR since it
represents two methods with different signatures. Restricting yourself in this
way also makes you think more carefully about the read vs. write behavior and
whether each should be part of the Public API.
.
.P
Here is an example TomDoc for \fBattr_reader\fR.
.
.IP "" 4
.
.nf

# Public: Get the user's name.
#
# Returns the String name of the user.
attr_reader :name
.
.fi
.
.IP "" 0
.
.P
Here is an example TomDoc for \fBattr_writer\fR. The parameter name should be the
same as the attribute name.
.
.IP "" 4
.
.nf

# Set the user's name.
#
# name \- The String name of the user.
#
# Returns nothing.
attr_writer :name
.
.fi
.
.IP "" 0
.
.P
While this approach certainly takes up more space than listing dozens of
attributes on a single line, it allows for individual documentation of each
attribute. Attributes are an extremely important part of a class and should be
treated with the same care as any other methods.


lib/tomdoc/tomdoc.rb
module TomDoc

  # TomDoc class needs Arg class.
  if RUBY_VERSION > '1.9'
    require_relative 'arg'
  else
    require 'tomdoc/arg'
  end

  class InvalidTomDoc < RuntimeError
    # Create new InvalidTomDoc object.
    #
    # doc - document string
    #
    def initialize(doc)
      @doc = doc
    end

    # Provide access to document string.
    #
    # Returns String.
    def message
      @doc
    end

    # Provide access to document string.
    #
    # Returns String.
    def to_s
      @doc
    end
  end

  #
  # TODO: Instead of having TomDoc::TomDoc, lets consider renaming this
  #       class TomDoc::Comment or something.
  #
  class TomDoc
    attr_accessor :raw

    # Public: Initialize a TomDoc object.
    #
    # text - The raw text of a method or class/module comment.
    #
    # Returns new TomDoc instance.
    def initialize(text, options={})
      @raw = text.to_s.strip

      @arguments        = []
      @examples         = []
      @returns          = []
      @raises           = []
      @signatures       = []
      @signature_fields = []

      parse unless @raw.empty?
    end

    def to_s
      @raw
    end

    # Validate given comment text.
    #
    # Returns true if comment is valid, otherwise false.
    def self.valid?(text)
      new(text).valid?
    end

    # Validate raw comment.
    #
    # Returns true if comment is valid, otherwise false.
    def valid?
      return false if !raw.include?('Returns')
      return false if sections.size < 2
      true
    end

    # Validate raw comment.
    #
    # Returns true if comment is valid.
    # Raises InvalidTomDoc if comment is not valid.
    def validate
      if !raw.include?('Returns')
        raise InvalidTomDoc.new("No `Returns' statement.")
      end

      if sections.size < 2
        raise InvalidTomDoc.new("No description section found.")
      end

      true
    end

    # The raw comment text cleaned-up and ready for section parsing.
    #
    # Returns cleaned-up comment String.
    def tomdoc
      lines = raw.split("\n")

      # remove remark symbol
      if lines.all?{ |line| /^\s*#/ =~ line }   
        lines = lines.map do |line|
          line =~ /^(\s*#)/ ? line.sub($1, '') : nil
        end
      end

      # for some reason the first line is coming in without indention
      # regardless, so we temporary remove it
      first = lines.shift

      # remove indention
      spaces = lines.map do |line|
        next if line.strip.empty?
        md = /^(\s*)/.match(line)
        md ? md[1].size : nil
      end.compact

      space = spaces.min || 0
      lines = lines.map do |line|
        if line.empty?
          line.strip
        else
          line[space..-1]
        end
      end

      # put first line back
      lines.unshift(first.sub(/^\s*/,''))

      lines.compact.join("\n")
    end

    # List of comment sections. These are divided simply on "\n\n".
    #
    # Returns Array of comment sections.
    def sections
      parsed {
        @sections
      }
    end

    # Description of method or class/module.
    #
    # Returns description String.
    def description
      parsed {
        @description
      }
    end

    # Description of method or class/module.
    #
    # Returns description String.
    def arguments
      parsed {
        @arguments
      }
    end
    alias args arguments

    # List of use examples of a method or class/module.
    #
    # Returns String of examples.
    def examples
      parsed {
        @examples
      }
    end

    # Description of a methods yield procedure.
    #
    # Returns String decription of yield procedure.
    def yields
      parsed {
        @yields
      }
    end

    # The list of retrun values a method can return.
    #
    # Returns Array of method return descriptions.
    def returns
      parsed {
        @returns
      }
    end

    # A list of errors a method might raise.
    #
    # Returns Array of method raises descriptions.
    def raises
      parsed {
        @raises
      }
    end

    # A list of alternate method signatures.
    #
    # Returns Array of signatures.
    def signatures
      parsed {
        @signatures 
      }
    end

    # A list of signature fields.
    #
    # Returns Array of field definitions.
    def signature_fields
      parsed {
        @signature_fields
      }
    end

    # Check if method is public.
    #
    # Returns true if method is public.
    def public?
      parsed {
        @status == 'Public'
      }
    end

    # Check if method is internal.
    #
    # Returns true if method is internal.
    def internal?
      parsed {
        @status == 'Internal'
      }
    end

    # Check if method is deprecated.
    #
    # Returns true if method is deprecated.
    def deprecated?
      parsed {
        @status == 'Deprecated'
      }
    end

  private

    # Has the comment been parsed yet?
    def parsed(&block)
      parse unless @parsed
      block.call
    end

    # Internal: Parse the Tomdoc formatted comment.
    #
    # Returns true if there was a comment to parse.
    def parse
      @parsed = true

      @sections = tomdoc.split("\n\n")
      sections  = @sections.dup

      return false if sections.empty?

      parse_description(sections.shift)

      if sections.first && sections.first =~ /^\w+\s+\-/m
        parse_arguments(sections.shift)
      end

      current = sections.shift
      while current
        case current
        when /^Examples/
          parse_examples(current, sections)
        when /^Yields/
          parse_yields(current)
        when /^(Returns|Raises)/
          parse_returns(current)
        when /^Signature/
          parse_signature(current, sections)
        end
        current = sections.shift
      end

      return @parsed
    end

    # Parse description.
    #
    # section - String containig description.
    #
    # Returns nothing.
    def parse_description(section)
      if md = /^([A-Z]\w+\:)/.match(section)
        @status      = md[1].chomp(':')
        @description = md.post_match.strip
      else
        @description = section.strip
      end   
    end

    # Parse arguments section. Arguments occur subsequent to
    # the description.
    #
    # section - String contaning agument definitions.
    #
    # Returns nothing.
    def parse_arguments(section)
      args = []
      last_indent = nil

      section.split("\n").each do |line|
        next if line.strip.empty?
        indent = line.scan(/^\s*/)[0].to_s.size

        if last_indent && indent > last_indent
          args.last.description += line.squeeze(" ")
        else
          param, desc = line.split(" - ")
          args << Arg.new(param.strip, desc.strip) if param && desc
        end

        last_indent = indent
      end

      @arguments = args
    end

    # Parse examples.
    #
    # section  - String starting with `Examples`.
    # sections - All sections subsequent to section.
    #
    # Returns nothing.
    def parse_examples(section, sections)
      examples = []

      section = section.sub('Examples', '').strip

      examples << section unless section.empty?
      while sections.first && sections.first !~ /^\S/
        examples << sections.shift.strip
      end

      @examples = examples
    end

    # Parse yields section.
    #
    # section - String contaning Yields line.
    #
    # Returns nothing.
    def parse_yields(section)
      @yields = section.strip
    end

    # Parse returns section.
    #
    # section - String contaning Returns and/or Raises lines.
    #
    # Returns nothing.
    def parse_returns(section)
      returns, raises, current = [], [], []

      lines = section.split("\n")  
      lines.each do |line|
        case line
        when /^Returns/
          returns << line
          current = returns
        when /^Raises/
          raises << line
          current = raises
        when /^\s+/
          current.last << line.squeeze(' ')
        else
          current << line  # TODO: What to do with non-compliant line?
        end
      end

      @returns, @raises = returns, raises
    end

    # Parse signature section.
    #
    # section  - String starting with `Signature`.
    # sections - All sections subsequent to section.
    #
    # Returns nothing.
    def parse_signature(section, sections=[])
      signatures = []

      section = section.sub('Signature', '').strip

      signatures << section unless section.empty?

      while sections.first && sections.first !~ /^\S/
        sigs = sections.shift
        sigs.split("\n").each do |s|
          signatures << s.strip
        end
      end

      @signatures = signatures

      if sections.first && sections.first =~ /^\w+\s*\-/m
        parse_signature_fields(sections.shift)
      end
    end

    # Subsequent to Signature section there can be field
    # definitions.
    #
    # section  - String subsequent to signatures.
    #
    # Returns nothing.
    def parse_signature_fields(section)
      args = []
      last_indent = nil

      section.split("\n").each do |line|
        next if line.strip.empty?
        indent = line.scan(/^\s*/)[0].to_s.size

        if last_indent && indent > last_indent
          args.last.description += line.squeeze(" ")
        else
          param, desc = line.split(" - ")
          args << Arg.new(param.strip, desc.strip) if param && desc
        end

        last_indent = indent
      end

      @signature_fields = args
    end

  end
end
